### PR TITLE
API Substitute silverstripe/config module for core config

### PR DESCRIPTION
**Pull request by tractorcow**
Implements https://github.com/silverstripe/silverstripe-framework/issues/6477

Config module work available for review at https://github.com/silverstripe/silverstripe-config/pull/4

Note that this change includes a few other api enhancements and changes designed to support the updated API. This includes a new module registration system (allowing modules to be referenced by full composer name), and changes to Director.

Non-trivial documentation updates will follow.

other PR:

 - [x] https://github.com/silverstripe/silverstripe-testsession/pull/45
 - [x] https://github.com/silverstripe/silverstripe-config/pull/4
 - [ ] https://github.com/silverstripe/silverstripe-graphql/pull/81
 - [ ] https://github.com/silverstripe/silverstripe-cms/pull/1739
 - [ ] https://github.com/silverstripe/silverstripe-asset-admin/pull/390
 - [ ] https://github.com/silverstripe/silverstripe-config/pull/5 (bugfix for infinity loop)
 - [ ] https://github.com/silverstripe/silverstripe-behat-extension/pull/148 (bugfix for missing `.env` handling)

### CODE REVIEW

File: src/Core/Config/CoreConfigCreator.php [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+use SilverStripe\Core\Config\Middleware\InheritanceMiddleware;
+use SilverStripe\Core\Manifest\ClassLoader;
+use SilverStripe\Core\Manifest\ModuleLoader;
+use Symfony\Component\Cache\Adapter\FilesystemAdapter;
+use Symfony\Component\Finder\Finder;
+
+/**
+ * Factory for silverstripe configs
+ */
+class CoreConfigCreator
```
Comment (chillu) [inline src/Core/Config/CoreConfigCreator.php]: What's the difference between a "factory" and a "creator"? We seem to use both terms in core

Comment (tractorcow) [inline src/Core/Config/CoreConfigCreator.php]: Synonyms.

File: src/Core/Config/CoreConfigCreator.php [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+     *
+     * @param bool $flush
+     * @return CachedConfigCollection
+     */
+    public function createRoot($flush)
+    {
+        $instance = new CachedConfigCollection();
+
+        // Set root cache
+        $instance->setPool(new FilesystemAdapter('configcache', 0, getTempFolder()));
```
Comment (chillu) [inline src/Core/Config/CoreConfigCreator.php]: Is there a plan for how to make this more pluggable? I guess it'll be easier once we create an `App` object as part of `main.php` bootstrap, where you could manually set services before relying on DI?

Comment (micmania1) [inline src/Core/Config/CoreConfigCreator.php]: Problem with that is the config is already built before it gets to the users code. You'd need to restructure how SS is bootstrapped.

Comment (micmania1) [inline src/Core/Config/CoreConfigCreator.php]: There's a little explanation and discussion here: https://github.com/silverstripe/silverstripe-framework/issues/6252#issuecomment-273914525

Comment (tractorcow) [inline src/Core/Config/CoreConfigCreator.php]: We do setup `.env` prior to this stage, so it's possible we could use non-yml/statics to configure this. maybe use `getenv`?

Comment (micmania1) [inline src/Core/Config/CoreConfigCreator.php]: The problem is that different backends have different constructors. eg. filesystem has none, redis has url, username, password etc.

Comment (sminnee) [inline src/Core/Config/CoreConfigCreator.php]: I recommend that we park config-cache-pluggability until after we have an app object.

File: src/Core/Config/CoreConfigCreator.php (hunk 1 above)
Comment (sminnee) [inline src/Core/Config/CoreConfigCreator.php]: Preference is to say "Factory"

File: src/Core/Config/Middleware/ExtensionMiddleware.php [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+{
+    use MiddlewareCommon;
+
+    /**
+     * Get config for a class
+     *
+     * @param string $class
+     * @param mixed $options
+     * @param callable $next
+     * @return string
```
Comment (sminnee) [inline src/Core/Config/Middleware/ExtensionMiddleware.php]: Return string|array|boolean|numeric I assume?

**Comment by sminnee** [comment]
@chillu Once https://github.com/silverstripe/silverstripe-config/pull/4 is merged, can you nudge the tests on this and assuming they pass merge it on Monday?

**Comment by tractorcow** [comment]
@sminnee I'll do a rebase and do the factory class rename at the same time.

**Comment by tractorcow** [comment]
Rebased and renamed that class.

**Comment by tractorcow** [comment]
Oops, rebase didn't remove Zend_Cache from new ModuleManifest. Fixed now. :D

**Comment by tractorcow** [comment]
Need a merge on https://github.com/silverstripe/silverstripe-config/pull/4 then we can restart this build.

**Comment by tractorcow** [comment]
Restarting with fix applied to testsession module

**Comment by tractorcow** [comment]
Please note that I've added a few "experimental" commits that should NOT be merged into the final PR. I'll remove these once I get tests green.

**Comment by tractorcow** [comment]
Added a new fix for https://github.com/silverstripe/silverstripe-config/pull/5

**Comment by tractorcow** [comment]
I've added two new bugfix PRs (see first comment) and pushed another change to this PR that should address certain issues.

**Comment by sminnee** [comment]
I've merged config#5 and re-triggered one of the builds on this PR to confirm that it doesn't break anything: https://travis-ci.org/silverstripe/silverstripe-framework/jobs/205966670#L210

**Comment by sminnee** [comment]
> Please note that I've added a few "experimental" commits that should NOT be merged into the final PR. I'll remove these once I get tests green.

@tractorcow can you confirm that these have now been removed and you're in principle happy to see this merged?

**Comment by tractorcow** [comment]
Yes those are removed. :)

**Pull request merged**